Data/Field.hs

```diff
@@ -0,0 +1,68 @@
+module Data.Field
+  ( -- * Field typeclass
+    Field(..)
+  , fromRational
+  ) where
+
+import Prelude hiding (fromRational, quot, (/))
+import Data.Complex (Complex)
+import Data.Euclidean (Euclidean, WrappedFractional, quot)
+import Data.Semiring (Ring, one)
+import Foreign.C.Types (CDouble, CFloat)
+import GHC.Real (Ratio((:%)))
+
+---------------------------------------------------------------------
+-- Classes
+---------------------------------------------------------------------
+
+-- | A field is a ring with a multiplicative inverse for any non-zero element.
+class (Euclidean a, Ring a) => Field a where
+
+  -- | Divide two elements of a 'Field'.
+  -- For any 'Prelude.Fractional' type, this is the same as '(Prelude./)'.
+  --
+  --     @x `divide` y = x `times` 'recip' y@
+  divide :: a -> a -> a
+  divide = quot
```

Comment:
The question is: are there any types for which the default definition of `divide` is not sufficient? If there are none, I would rather move `divide` and `recip` out: 

```haskell
class (Euclidean a, Ring a) => Field a 

divide :: Field a => a -> a -> a 
divide = quot

recip :: Field a => a -> a -> a 
recip = quot one 
```

Reply:
Is there a general convention for this sort of thing, or if having extraneous functions in a class would affect performance? It's just my personal preference for `divide` to be part of the class, just like how `plus` and `times` are, as not having `minus` in `Ring` was slightly annoying for me (e.g. when doing explicit imports).

Reply:
Type classes are represented in run time as records of functions. Additional type class methods inflate these records. It is difficult to say how severe is the impact on performance in this case, but it is not free. I've seen examples where passing records around took more time than actual computations.

Reply:
Classes with 0 functions are equivalent to () at runtime.
Classes with 1 function are equivalent to a newtype with a single function at runtime.
Classes with 2+ functions are equivalent to a datatype.

The first two cases are more efficient in general, with two caveats:
  1. This of course ignores the superclass constraints, but that can be viewed as constant overhead that affects all three cases.
  2. This also ignores the possibility of specialisation.

So, it's better to be on the safe side and factor out unnecessary bits. Note that in `Ring` we only have one function, `negate`, where we would also probably do well to have `minus`. The reasoning behind our choice is what I've talked about above, and we assume there's no better implementation for `minus` in most cases.

Reply:
I see - thanks.